### ISSUE TITLE
Questions on Mapping map<text,?>

### ISSUE THREAD

**Issue by kyleboyle** [strapdata/elassandra#92]
1. Is it possible to have elassandra translate c* maps to type `object` and not `nested`. I'd like to be able to specify filters without defining a nested query, which is required by the `nested` field type. This also would allow using the URI query like `q=mapname.key=*searchvalue*`. I tried manually setting the type to `object` but it didn't seem to have an effect on functionality, and the `type:object` wasn't persisted in the metadata for the field.

Ref: [inner objects](https://www.elastic.co/guide/en/elasticsearch/guide/current/complex-core-fields.html#inner-objects)     [nested objects](https://www.elastic.co/guide/en/elasticsearch/guide/current/nested-objects.html#nested-objects)

2. Is it possible to specify indexing properties on the object sub fields. By default, only the data type `type` is added. For example I'd like the following:
```
{
  "event_test_index" : {
    "aliases" : { },
    "mappings" : {
      "event_test" : {
        "properties" : {
          "id" : {
            "type" : "string",
            "index" : "not_analyzed",
            "cql_collection" : "singleton",
            "cql_partition_key" : true,
            "cql_primary_key_order" : 0
          },
          "strings" : {
            "cql_struct" : "map",
            "type" : "nested",
            "cql_collection" : "singleton",
            "properties" : {
              "stringProp1" : {
                "type" : "string"
              },
              "stringProp2" : {
                "type" : "string"
              }
            }
          }
        }
      }
    },
    "settings" : {
      "index" : {
        "keyspace" : "keyspace1",
        "creation_date" : "1491912302498",
        "number_of_shards" : "1",
        "number_of_replicas" : "0",
        "uuid" : "XkKpUp1zRzmD1eofD4xCIQ",
        "version" : {
          "created" : "2040299"
        }
      }
    },
    "warmers" : { }
  }
}
```
to have additional properties for the object fields (keys) as they are added to the mapping as new data is indexed:
```
{
  "event_test_index" : {
    "aliases" : { },
    "mappings" : {
      "event_test" : {
        "properties" : {
          "id" : {
            "type" : "string",
            "index" : "not_analyzed",
            "cql_collection" : "singleton",
            "cql_partition_key" : true,
            "cql_primary_key_order" : 0
          },
          "strings" : {
            "cql_struct" : "map",
            "type":"nested",
            "cql_collection" : "singleton",
            "properties" : {
              "stringProp1" : {
                "type" : "string",
                "doc_values": true,                       // <- additional index settings
                "index": "not_analyzed"                   // <- additional index settings
              },
              "stringProp2" : {
                "type" : "string",
                "doc_values": true,                       // <- additional index settings
                "index": "not_analyzed"                   // <- additional index settings
              }
            }
          }
        }
      }
    },
    "settings" : {
      "index" : {
        "keyspace" : "keyspace1",
        "creation_date" : "1491912302498",
        "number_of_shards" : "1",
        "number_of_replicas" : "0",
        "uuid" : "XkKpUp1zRzmD1eofD4xCIQ",
        "version" : {
          "created" : "2040299"
        }
      }
    },
    "warmers" : { }
  }
}
```

Thank you

**Comment by kyleboyle** [comment]
To answer my own question number 1, it appears that if I manually set the map type to 'object', I can search without using nested queries. I was thrown off because the type isn't persisted as 'object' in the metadata (mapping) - not sure if this is by design. eg I can do this query:
`curl -GET "http://localhost:9200/event_test_index/event_test/_search?q=strings.stringProp4:*value*&pretty=true"`

Still curious on your response to number 2

Thanks

**Comment by vroyer** [comment]
Well, for question 1, you could add include_in_root:true (or include_in_parent:true) in the nested object mapping. Then, you can search on nested fields without using a nested query.

Currently, dynamic mapping of C* map does not use ES template to automatically add include_in_root or any additional settings, you have to provide this mapping for each key in your underlying C* map.  This could probably be added in the futur.

```
curl -XPUT "http://$NODE:9200/test" -d '{
  "mappings": {
      "event_test" : {
            "properties" : {
              "strings" : {
                "cql_struct" : "map",
                "cql_collection" : "singleton",
                "include_in_root": true,
                "type" : "nested",
                "properties" : {
                  "r2" : {
                    "type" : "string"
                  }
                }
              },
              "id" : {
                "index" : "not_analyzed",
                "cql_partition_key" : true,
                "cql_primary_key_order" : 0,
                "type" : "string",
                "cql_collection" : "singleton"
              }
            }
          }
  }
}'
```

cqlsh> insert into test.event_test (id,strings) VALUES ('3',{'r2':'b'});

```
get 'test/_search?pretty&q=strings.r2:b'
{
  "took" : 9,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "failed" : 0
  },
  "hits" : {
    "total" : 1,
    "max_score" : 0.5945348,
    "hits" : [ {
      "_index" : "test",
      "_type" : "event_test",
      "_id" : "3",
      "_score" : 0.5945348,
      "_source" : {
        "strings" : {
          "r2" : "b"
        },
        "id" : "3"
      }
    } ]
  }
}
```

**Comment by kyleboyle** [comment]
Thanks Vincent. Could you comment on the behaviour of explicitly setting the map's type to `object` rather `nested`. Indexing appears to work as if the type is object (ie not a nested doc), but the type isn't returned in the index mapping metadata:

```
curl -XPUT "http://localhost:9200/event_test_index" -d '{
>   "settings": {
>       "keyspace":"keyspace1"
>   },
>   "mappings": {
>       "event_test" : {
>
> "properties": {
>           "id": {
>             "type": "string",
>             "index": "not_analyzed",
>             "cql_collection": "singleton",
>             "cql_partition_key": true,
>             "cql_primary_key_order": 0
>           },
>           "strings": {
>             "type": "object",
>             "cql_struct": "map",
>             "cql_collection": "singleton"
>           }
>         }
>
>
>        }
>   }
> }'
```

```
curl -XGET "http://localhost:9200/event_test_index?pretty"
{
  "event_test_index" : {
    "aliases" : { },
    "mappings" : {
      "event_test" : {
        "properties" : {
          "id" : {
            "type" : "string",
            "index" : "not_analyzed",
            "cql_collection" : "singleton",
            "cql_partition_key" : true,
            "cql_primary_key_order" : 0
          },
          "strings" : {                      // missing "type":"object"
            "cql_struct" : "map",
            "cql_collection" : "singleton",
            "properties" : {
              "stringProp3" : {
                "type" : "string"
              },
              "stringProp4" : {
                "type" : "string"
              }
            }
          }
        }
      }
    },
    "settings" : {
      "index" : {
        "keyspace" : "keyspace1",
        "creation_date" : "1492015248367",
        "number_of_shards" : "1",
        "number_of_replicas" : "0",
        "uuid" : "bvz0EqaNSfyiOeLW4uG7MQ",
        "version" : {
          "created" : "2040299"
        }
      }
    },
    "warmers" : { }
  }
}
```

**Comment by vroyer** [comment]
Yes, this is the elasticsearch standard behavior, "object" is the default for and object having some nested fields, so it is not displayed as soon as you have at least one sub field. Here is the output of elasticsearch 2.4.1 with such a mapping : 

```
{
  "cluster_name" : "elasticsearch",
  "metadata" : {
    "cluster_uuid" : "f7yvvHWJReCX11xAVk2rHg",
    "templates" : { },
    "indices" : {
      "event_test_index" : {
        "state" : "open",
        "settings" : {
          "index" : {
            "creation_date" : "1492025549017",
            "number_of_shards" : "5",
            "number_of_replicas" : "1",
            "uuid" : "q6dZnA3HQpyf5oymYSw1VQ",
            "version" : {
              "created" : "2040199"
            }
          }
        },
        "mappings" : {
          "event_test" : {
            "properties" : {
              "strings" : {
                "properties" : {
                  "t" : {
                    "type" : "string"
                  }
                }
              },
              "id" : {
                "index" : "not_analyzed",
                "type" : "string"
              }
            }
          }
        },
        "aliases" : [ ]
      }
    }
  }
}
```

**Comment by kyleboyle** [comment]
Perfect, thank you.

Regarding functionality required to handle explicit settings for dynamically created fields, would your preferred implementation aim to provide [dynamic template](https://www.elastic.co/guide/en/elasticsearch/guide/2.x/custom-dynamic-mapping.html#dynamic-templates) functionality? In the vein of my previous example:

```

    "mappings" : {
      "event_test" : {
           "dynamic_templates": [
                { "strings_template": {
                      "match": "strings.*", 
                      "mapping": {
                          "type": "string",
                          "doc_values": true,
                          "index": "not_analyzed"
                      }
                }}
            ],
        "properties" : {
          "id" : {
            "type" : "string",
            "index" : "not_analyzed",
            "cql_collection" : "singleton",
            "cql_partition_key" : true,
            "cql_primary_key_order" : 0
          },
          "strings" : {
            "cql_struct" : "map",
            "type" : "object",
            "cql_collection" : "singleton"
          }
        }
      }
    }
```

**Comment by vroyer** [comment]
This will be included in the next release for next week. 
We also refactor the source code to loosely couple source code from cassandra and elasticsearch, making contribution easier.

**Comment by vroyer** [comment]
Fixed in release 2.4.2.13+
